flashlight/fl/tensor/backend/jit/ir/Node.cpp

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "flashlight/fl/tensor/backend/jit/ir/Node.h"
+
+#include <cassert>
+#include <sstream>
+#include <stdexcept>
+
+namespace fl {
+
+void Node::nodeImplTypeCheck(NodeType expect, NodeType actual) const {
+  if (expect != actual) {
+    std::ostringstream oss;
+    oss << "[fl::Node::impl] "
+        << "specified node type: [" << actual << "] "
+        << "doesn't match actual node type: [" << expect << "]";
+    throw std::invalid_argument(oss.str());
+  }
+}
+
+void Node::setInputImpl(unsigned inputIdx, Node* input) {
+  inputs_.at(inputIdx) = input;
+  // update metadata
+  auto inputUse = Use::create(this, inputIdx);
+  auto inputUseIter = input->uses_.insert(input->uses_.end(), inputUse);
+  inputUseIters_[inputIdx] = inputUseIter;
+  input->incRefCount();
+}
+
+void Node::resetInput(unsigned inputIdx) {
+  const auto oldInput = inputs_.at(inputIdx);
+  const auto oldInputUseIter = inputUseIters_[inputIdx];
+  assert(oldInput && "No input at given index");
+  // reset
+  inputUseIters_[inputIdx] = oldInput->uses_.end();
+  inputs_[inputIdx] = nullptr;
+  // clean up metadata
+  delete *oldInputUseIter;
+  oldInput->uses_.erase(oldInputUseIter);
+  oldInput->decRefCount();
+}
+
+Node::Node(std::vector<Node*>&& inputs) {
```

Comment:
you may want to calculate a quick `hash` here :)

Reply:
Good point. I was thinking about that, but decided to keep it simple for now -- until hashing becomes needed later (e.g., for caching). Nice thing is that it shouldn't be hard to add.